aio.core/aio/core/directory/abstract/directory.py

```diff
@@ -0,0 +1,203 @@
+import pathlib
+import shutil
+from functools import cached_property, lru_cache
+from typing import (
+    AsyncIterator, Iterable, Optional,
+    Pattern, Set, Tuple, Union)
+
+import abstracts
+
+from aio.core import subprocess
+from aio.core.functional import async_property, async_set, AwaitableGenerator
+from aio.core.directory import exceptions
+
+
+class ADirectory(metaclass=abstracts.Abstraction):
+    """A filesystem directory, with an associated fileset, and tools for
+    finding files.
+
+    The required `path` argument should be the path to an existing filesystem
+    directory.
+
+    The `exclude` and `exclude_dirs` parameters are passed to the `grep`
+    command as `--exclude` and `--exclude-dir` args respectively.
+
+    You can specify inclusive and exclusive regex matchers with `path_matcher`
+    and `exclude_matcher`.
+
+    By default, only text files (from grep/git grep pov) will be searched,
+    you can override this by setting `text_only` to `False`.
+    """
+
+    def __init__(
+            self,
+            path: Union[str, pathlib.Path],
+            exclude: Optional[Iterable[str]] = None,
+            exclude_dirs: Optional[Iterable[str]] = None,
+            path_matcher: Optional[Pattern[str]] = None,
+            exclude_matcher: Optional[Pattern[str]] = None,
+            text_only: Optional[bool] = True) -> None:
+        self._path = path
+        self.path_matcher = path_matcher
+        self.exclude_matcher = exclude_matcher
+        self.text_only = text_only
+        self.exclude = exclude or ()
+        self.exclude_dirs = exclude_dirs or ()
+
+    @async_property(cache=True)
+    async def files(self) -> Set[str]:
+        """Set of relative file paths associated with this directory."""
+        return await self.grep(["-l", ""])
+
+    @cached_property
+    def grep_args(self) -> Tuple[str, ...]:
+        """Default args to pass when calling `grep`."""
+        return (
+            *self.grep_command,
+            *(("-I", )
+              if self.text_only
+              else ()),
+            *self.grep_exclusions)
+
+    @property
+    def grep_command(self) -> Tuple[str, ...]:
+        """Path to the `grep` command."""
```

Comment:
this is a tuple rather than just a path - probs change the name, and fix the comment

Reply:
`grep_command_args`